EIPS/eip-4804.md

```diff
@@ -0,0 +1,140 @@
+- **contractName** indicates the contract to be called, i.e., "To" field in the EVM call message. If the **contractName** is an **address**, i.e., 0x + 20-byte-data hex, then "To" will be the address. Otherwise, the name is from a name service. In the second case, **nsProvider** will be the short name of name service providers such as "ens", "w3q", etc. The way to translate the name from a name service to an address will be defined in future proposals.
+- **chainid** indicates which chain to call the message. If not specified, the protocol will use the same chain as the name service provider, e.g., 1 for eth, and 333 for w3q. If no name service provider is available, the default chainid is 1.
+- **returnTypes** tells the format of the returned data. If not specified, the returned message data will be parsed in "(bytes32)" and MIME will be set based on the suffix of the last argument. If **returnTypes** is "()", the returned data will be parsed in raw bytes in JSON.  Otherwise, the returned message will be parsed in the specified "returnTypes" in JSON.  
+
+### Resolve Mode
+
+Once the "To" address and chainid are determined, the protocol will check the resolver mode of contract by calling "resolveMode" method. The protocol currently supports two resolve modes:
+
+#### Manual Mode
+
+The manual mode will not do any interpretation of **path**, and put **path** as calldata of the message directly.
+
+#### Auto Mode
+
+The auto mode is the default mode of resolver (also applies when the "resolveMode" method is unavailable in the target contract). In the auto mode, if **path** is empty, then the protocol will call the target contract with empty calldata. Otherwise, the calldata of the EVM message will use standard Ethereum contract ABI [Contract ABI Specification — Solidity 0.8.3 documentation](https://docs.soliditylang.org/en/v0.8.3/abi-spec.html), where
+
+- **method** is a string of function method be called
+- **argument_i** is the ith argument of the method. If **type** is specified, the value will be translated to the corresponding type. The protocol currently supports the basic types such as uint256, bytes32, address, and bytes. If **type** is not specified, then the type will be automatically detected using the following rule in a sequential way:
+
+1. **type** = "uint256", if **value** is numeric; or
+2. **type**="bytes32", if **value** is in the form of 0x+32-byte-data hex; or
+3. **type**="address", if **value** is in the form of 0x+20-byte-data hex; or
+4. **type**="name", if **value** is in the form of **name**.**nsProvider**. In this case, the actual value of the argument will be **address**, which is obtained from **nsProvider**, e.g., eth, w3q, etc.
+5. else **type**="bytes"
+
+Note that if **method** does not exist, i.e., **path** is empty or "/", then the contract will be called with empty calldata.
+
+## Examples
```

Comment:
This is not a valid top level section according to [EIP-1](https://eips.ethereum.org/EIPS/eip-1).

Reply:
Changed to ### Examples